syspage.c

```diff
@@ -479,42 +535,51 @@ int syspage_addProg(void *start, void *end, const char *imap, const char *dmap,
 	syspage_common.args[syspage_common.argCnt] = '\0';
 
 	/* copy only program name, without (;) args) */
-	low_memcpy(syspage_common.progs[progID].cmdline, name, pos);
+	low_memcpy(prog->cmdline, name, pos);
 
 	while (pos < MAX_CMDLINE_SIZE)
-		syspage_common.progs[progID].cmdline[pos++] = '\0';
+		prog->cmdline[pos++] = '\0';
 
 	syspage_common.progsCnt++;
 
-	return 0;
+	return ERR_NONE;
 }
 
 
 
 /* Setting kernel's data */
 
-void syspage_setKernelText(void *addr, u32 size)
+void syspage_setKernelText(void *addr, size_t size)
 {
-	syspage_common.syspage->kernel.text = addr;
+	syspage_common.syspage->kernel.text = (addr_t)addr;
 	syspage_common.syspage->kernel.textsz = size;
 
-	syspage_addEntries((u32)addr, size);
+	syspage_addEntries((addr_t)addr, size);
 }
 
 
-void syspage_setKernelBss(void *addr, u32 size)
+void syspage_setKernelBss(void *addr, size_t size)
 {
-	syspage_common.syspage->kernel.bss = addr;
+	syspage_common.syspage->kernel.bss = (addr_t)addr;
 	syspage_common.syspage->kernel.bsssz = size;
 
-	syspage_addEntries((u32)addr, size);
+	syspage_addEntries((addr_t)addr, size);
 }
 
 
-void syspage_setKernelData(void *addr, u32 size)
+void syspage_setKernelData(void *addr, size_t size)
 {
-	syspage_common.syspage->kernel.data = addr;
+	syspage_common.syspage->kernel.data = (addr_t)addr;
 	syspage_common.syspage->kernel.datasz = size;
 
-	syspage_addEntries((u32)addr, size);
+	syspage_addEntries((addr_t)addr, size);
 }
+
+
+/* Add specific hal data */
+
+void syspage_setHalData(const syspage_hal_t *hal)
+{
+	low_memcpy((void *)&syspage_common.hal, (void *)hal, sizeof(syspage_hal_t));
+}
+
```

Comment:
In the `syspage_save()` function you check the condition `sizeof(syspage_hal_t) != 0`, here you don't.
I suggest adding this condition here ( it will be a bit shorter and faster code ).
I also suggest you don't casting pointers to `(void *)` when calling `low_memcpy()`, it's unnecessary and e.g. not used in `phoenix-rtos-kernel` in `hal_memcpy()` call.